### ISSUE TITLE
Terrain shader plans

### ISSUE THREAD

**Issue by BlackYps** [FAForever/fa#4906]
Terrain shading looks rather dated and could use an update. For this we need to make a plan what we want to change in the shader and how this would affect the inputs and the workflow for mappers.

Currently the stratum masks only work in the upper half. They work in the range 0.5..1 instead of 0..1. Mappers use shadow and normal decals to alleviate the current shortcomings, but it's not a satisfying solution.

At the moment the map shader has the following inputs:
```
             | Albedo stratum                                     | Normal stratum
|Layer | R        | G            | B           | A              | R            | G          | B     | A             |
---------------------------------------------------------------------------------------------------------------------
|lower |        R            G               B            unused     | X               Y               Z       unused

|0 |        R            G               B            unused     | X               Y               Z       unused
|1 |        R            G               B            unused     | X               Y               Z       unused
|2 |        R            G               B            unused     | X               Y               Z       unused
|3 |        R            G               B            unused     | X               Y               Z       unused

|4 |        R            G               B            unused     | X               Y               Z       unused
|5 |        R            G               B            unused     | X               Y               Z       unused
|6 |        R            G               B            unused     | X               Y               Z       unused
|7 |        R            G               B            unused     | X               Y               Z       unused

|upper |        R            G               B             alpha
```

Each layer has texture scale values for the albedo and normal maps.
Aditionally we have two mask textures that store the opacity of layers 0-7 in their channels.

---------

We can repurpose stratum 7 and the upper stratum to fill them with our own inputs. This gives is new input channels to control the terrain shader. This additionaly frees up one mask slot because we don't need a mask for stratum 7 anymore.

These free channels could be used for different things:
terrain shadows (1 channel) - This way we don't need to use a shadow decal anymore which is prone to flickering.
mapwide normal (2 channels) - This way we don't need to use a normal decal anymore. Apart from getting rid of the flickering, this allows proper interpolation with the underlying terrain.
mapwide albedo (either 1 or 3 channels) - this allows to bring in detail that is visible when being zoomed out. With one channel we could do a simple lightening/darkening of the albedo, with three channels we could tune the color as well.
mask for changing texture sample direction to prevent tiling (1 channel)
specularity/roughness (1 channel) - could be used to give shiny surfaces like ice or mud sharper reflections. We only have the sun reflection to work with, so it won't look amazing, but it's better than nothing
bloom factor (1 channel) - don't really know how useful that is, but could be interesting

Things that are not feasible to implement:
metal shader - we don't have access to an environment map. As metal relies a lot on reflections to look convincing, there is no way to make it work with the constraints we have.
ambient occlusion - we can't access (or at least I don't know of a way) these channels in the mesh shader, so props and units would not know the ambient occlusion lighting information which would make lighting inconsistent with the terrain.

short-term plans
-----
We could easily make the masks work on the full range.
We can implement the shadow and mapwide normal channels.
This leaves one texture slot free for a macro texture.
We could fit all this into the upper layer which would keep the 8 stratum layers.

Providing this under a new shader name enables mappers to choose which shader they want to use. Adjusting existing maps with mapwide decals would be a relatively easy process, because mappers don't have to come up with solutions for one stratum missing that they potentially have used.

long-term plans
----
the albedo and normal scaling are independent of each other at the moment. It doesn't really make sense to provide this option, because it is good practice to use albedos with corresponding normal maps. By unifying the scaling of both we free up one scalar value per stratum. We could use this to control the scaling of a second sampling of the textures to provide more perceived resolution and to break up texture repetition.
If we notice that the second sampling should ideally happen at a fixed factor compared to the original scaling, then we could use this value to control the normal strenght of the second pass instead.

We could require that all stratum masks always add up to 1, this way we can utilize more advanced texture splatting techniques like this: https://www.gamedeveloper.com/programming/advanced-terrain-texture-splatting

We can also switch to better normal interpolation between the layers.

These changes would require a map visualizer or editor that is capable of displaying the new shader, to enable mappers to not work blindly on their maps and relying on the game to visualize the map, which is far too tedious.

**Comment by BlackYps** [comment]
Another idea: We need some sort of height information for the textures to really make use of the advanced texture splatting. The alpha channel of the textures is unused at the moment, so an obvious first idea is to use that. However, this would require us to process all of the textures. It also makes using custom textures harder, because you first have to edit their alpha channel. Instead we could use our free channels to provide different height textures. This way we don't have to edit all of the textures and we can even switch out height textures easily to try out different height profiles. If we combine four of them as a texture atlas then we only need two channels to provide eight different height profiles.

**Comment by BlackYps** [comment]
Here is my proposal for the texture usage:

upperAlbedo -- map size * (1 or 2)
 R: normal.x
 G: normal.z
 B: shadow
 A: albedo overlay

Stratum7Albedo - texture size * 2
 R: height stratum 0-3
 G: roughness stratum 0-3
 B: height stratum 4-7
 A: roughness stratum 4-7

Stratum7Normal -- environmentMap size
 R: environmentMap.r
 G: environmentMap.g
 B: environmentMap.b
 A: unused, but not easily usable anyway

mask1.w -- map size
 texture sample direction or roughness overlay

stratum8AlbedoScale
 normal scale factor for near pass

stratum8NormalScale
 normal scale factor for far pass
 
Some thoughts on this
------

We do two texture samples and the stratum albedo scales would control the near texture scale and the stratum albedo scales control the far scale.
The stratum8 scales globally control the intensity of the normal textures. These should offer a 0..2 range and default to 1 in the editor

Changing the sample direction potentially hugely increases the complexity in the shader, because you casually double the number of sampled textures you have to store. So it may be more sensible to use the mask1.w channel for a roughness overlay instead, changing the roughness of the textures at that location. The roughness and albedo overlay can then be used in combination to create a wetness effect on the surface.

Adding the environment map would allow us create better reflections for materials with low roughness. It would also allow us to truely match the lighting of the PBR unit shader with the terrain.
If we break compatibility with the legacy lighting (or create a second version of the PBR unit shader) then we could hugely simplify the lighting process in both the mesh and terrain shader and potentially free a map-wide rgb value, though I wouldn't really know what to do with it.

Instead of using an environment map, we could instead use three channels to basically create a map-wide albedo decal. Though I don't think I have heard of such an albedo decal being used, so it's probably not that interesting to have. But I am no mapper, so it would be nice to have another opinion on this.

PBR:
PBR reflections based on roughness could be tricky, we either need the brdf lookup texture which uses two slots or we have to approximate the values with functions.
Metal is a bit gimmicky, there is basically no metal in natural environments so it would be completely wasted on most maps and would use a lot of texture slots.

**Comment by Garanas** [comment]
We should start simple and just implement this bit:

```
upperAlbedo -- map size * (1 or 2)
 R: normal.x
 G: normal.z
 B: shadow
 A: -
 ```
 
Introducing these few channels would be a significant task already. Luckily the map generator already generates these channels. We'd need to adjust where these channels end up. We'd remove the corresponding decals and instead create a texture that we can use for the upper stratum.  

Once that is done, we can look at edge cases. Take albedo decals as an example. How do those sample the shadow map? Something I think we could do is store the shadow map in the render phase where we compute the normals. (as one of the channels, there are 2 free channels in the end). Then we can sample that along with the normals, and apply it the same way we do in the terrain shader. This would also require us to change how normal decals work. And we'd need to make the channels we use for shadows entirely white for the other render phases, as the decal shaders are shared across maps.

And at last we end at the tweaking phase, and the question how we take into account the shadow color. And do we consider the map-wide normal to be in local space (where it adds details to the normals of the terrain) or do we consider it to be in world space, where we just replace the normals of the terrain? I'm in favor of them being in world space. And I bet we'll run into more issues as we implement these three channels.

As @BlackYps , these would replace map-wide decals. Especially the flickering is annoying. We partially fix that in #4915 , but it would be great to be able to fix that for good for these crucial channels.

**Comment by BlackYps** [comment]
Sheikah and me already  coded the necessary changes to the map generator, so it outputs the required channels into a texture that gets loaded as the upper albedo.
The shadow color is taken into account by the terrain shader. The shadow hannel only stores where shadow is supposed to be. Similar to the shadow map that the game produces dynamically. By multiplying both together we can then feed this shadow information into the shader as normal and the shadow color will be applied correctly.
Do I understand this correctly that you suggest to use the normal channel, so we don't have to sample the upper stratum in the decal shader as that would be the same shader for all maps and would therefore destroy the look on old maps that use a regular upper albedo texture? That approach would work in this specific case, but would break down at the latest when we want to change the specular response.
Couldn't we use a trigger like a certain upper albedo scale to decide between two different implementations of the shader? We don't need this scale anymore, so we could create an if statement that triggers when the scale is 0. Or something along that.
I thought that one advantage of using the normal information in the stratum is that we can choose other interpolation options than just replacing the normals of the terrain. This way we can get away with lower resolution normal maps as the details from the stratum layers is still shining through, hiding the lower resolution of our custom map-wide normals. Is that not so? What would be the advantage of leaving them in world space? And isn't local vs world space something that can be transformed with a simple matrix multiplication anyway?

**Comment by BlackYps** [comment]
This project is implemented. Documentation can be found here: https://wiki.faforever.com/en/Development/Shaders/terrain-shaders
